### PR TITLE
Support DCN parallelism

### PR DESCRIPTION
**Pull request by jonb377**
- Add `spmd_dcn_parallelism` to define dcn parallelism.
- Reuse a single mesh across all sharding
- Shard parameters in `__init__` to avoid host-side OOM

### CODE REVIEW

File: examples/pytorch/language-modeling/run_clm.py [hunk 1]
```diff
@@ -543,35 +551,51 @@ def get_mesh(ici_mesh_shape, dcn_mesh_shape=None):
...
           mesh_shape = [1] * len(param.shape)
           mesh_shape[max_dim] = num_devices
-          print('> [FSDP] Sharding tensor', name, param.shape, mesh_shape)
-          mesh = get_mesh(tuple(mesh_shape))
-          xs.mark_sharding(param, mesh, range(len(param.shape)))
+
+          # Place DCN parallelism on an independent axis
+          dcn_mesh_shape = (model_args.spmd_dcn_parallelism,) + (1,) * len(param.shape)
+          ici_mesh_shape = (1,) + tuple(mesh_shape)
+          mesh = get_mesh(ici_mesh_shape, dcn_mesh_shape)
```
Comment (JackCaoG) [inline examples/pytorch/language-modeling/run_clm.py]: what will get_mesh return here?

Comment (jonb377) [reply examples/pytorch/language-modeling/run_clm.py]: Ah, I need to cut the FSDP dim by `dcn_parallelism`. The result should be `(dcn, n_dev // dcn, 1)` if max_dim = 0

Comment (JackCaoG) [reply examples/pytorch/language-modeling/run_clm.py]: hmm, what does this `get_mesh` function and why do we need to pass both `ici` and `dcn` mesh shapes?

Comment (jonb377) [reply examples/pytorch/language-modeling/run_clm.py]: `get_mesh` allows us to switch between the iota mesh and HybridMesh. HybridMesh requires both `ici` and `dcn` shapes as input, so we need to keep them separate here as well

Comment (jonb377) [reply examples/pytorch/language-modeling/run_clm.py]: The return is either `xs.Mesh` or `xs.HybridMesh` depending on the `spmd_iota_mesh` flag

Comment (JackCaoG) [reply examples/pytorch/language-modeling/run_clm.py]: ah nice

**Comment by jonb377** [comment]
Converting to draft for now, I need to update activation sharding as well

**Comment by alanwaketan** [comment]
Okay, give me one moment before my PR is ready. Can you land #26? And I'm going to rebase on it.

**Comment by jonb377** [comment]
> Okay, give me one moment before my PR is ready. Can you land #26? And I'm going to rebase on it.

Done, thanks @alanwaketan

**Review by JackCaoG — APPROVED** [review]

**Review by alanwaketan — APPROVED** [review]
LGTM, but let's us hold this until we finalize all the changes needed for Next.

File: src/transformers/models/llama/modeling_llama.py [hunk 2]
```diff
@@ -195,39 +195,34 @@ def apply_rotary_pos_emb(q, k, cos, sin, position_ids):
...
-        self.gate_proj = nn.Linear(self.hidden_size, self.intermediate_size, bias=False)
-        self.up_proj = nn.Linear(self.hidden_size, self.intermediate_size, bias=False)
-        self.down_proj = nn.Linear(self.intermediate_size, self.hidden_size, bias=False)
+        self.gate_proj = nn.Linear(self.hidden_size, self.intermediate_size, bias=False).to('xla')
+        self.up_proj = nn.Linear(self.hidden_size, self.intermediate_size, bias=False).to('xla')
+        self.down_proj = nn.Linear(self.intermediate_size, self.hidden_size, bias=False).to('xla')
         self.act_fn = ACT2FN[config.hidden_act]
 
-        init_spmd(self, config)
+        if self.spmd_fsdp_sharding:
```
Comment (alanwaketan) [inline src/transformers/models/llama/modeling_llama.py:217]: Are we dropping tensor sharding?

File: src/transformers/models/llama/modeling_llama.py [hunk 3]
```diff
@@ -195,39 +195,34 @@ def apply_rotary_pos_emb(q, k, cos, sin, position_ids):
...
+            gate_proj_spec = up_proj_spec = down_proj_spec = ('data', None)
+        else:
```
Comment (alanwaketan) [inline src/transformers/models/llama/modeling_llama.py:219]: Can we comment that this is for 2D sharding?

File: src/transformers/models/llama/modeling_llama.py [hunk 4]
```diff
@@ -312,13 +309,22 @@ def __init__(self, config: LlamaConfig):
...
-        self.v_proj = nn.Linear(self.hidden_size, self.num_key_value_heads * self.head_dim, bias=False)
-        self.o_proj = nn.Linear(self.num_heads * self.head_dim, self.hidden_size, bias=False)
+        self.q_proj = nn.Linear(self.hidden_size, self.num_heads * self.head_dim, bias=False).to('xla')
+        self.k_proj = nn.Linear(self.hidden_size, self.num_key_value_heads * self.head_dim, bias=False).to('xla')
+        self.v_proj = nn.Linear(self.hidden_size, self.num_key_value_heads * self.head_dim, bias=False).to('xla')
+        self.o_proj = nn.Linear(self.num_heads * self.head_dim, self.hidden_size, bias=False).to('xla')
         self._init_rope()
 
-        init_spmd(self, config)
+        if self.spmd_fsdp_sharding:
```
Comment (alanwaketan) [inline src/transformers/models/llama/modeling_llama.py:319]: Same comments as above.

File: src/transformers/models/llama/modeling_llama.py (hunk 2 above)
Comment (alanwaketan) [reply src/transformers/models/llama/modeling_llama.py:217]: I feel like we should drop it as it's not useful anyway and then it's not compatible with the 2D shading scheme. Let me do a follow up to drop it later.

**Review by alanwaketan — APPROVED** [review]
LGTM.

File: src/transformers/models/llama/modeling_llama.py (hunk 3 above)
Comment (jonb377) [reply src/transformers/models/llama/modeling_llama.py:219]: Sure, I'll add some more comments throughout
